File: UNIT-01/orchestrator/auto_context.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"unit01/clients"
)

// getAutoContext uses Indexer's semantic/fuzzy search to pre-fetch context based on the raw user input.
func getAutoContext(input string, ws *Workspace) string {
	if !ws.Active {
		return ""
	}

	indexer := clients.NewIndexerClient()
	records, err := indexer.Search(input)
	if err != nil || len(records) == 0 {
		if strings.Contains(strings.ToLower(input), "content") || strings.Contains(strings.ToLower(input), "list") || strings.Contains(strings.ToLower(input), "folder") {
			list, err := indexer.List(ws.Path)
			if err == nil && list != nil && len(list.Entries) > 0 {
				var context strings.Builder
				context.WriteString("\n# DIRECTORY CONTENTS:\n")
				for _, item := range list.Entries {
					context.WriteString(fmt.Sprintf("- [%s] %s\n", item.Type, item.Name))
				}
				return context.String()
			}
		}

		// FALLBACK: Search chat history for relevant past discussions
		chatHistoryCtx := searchChatHistory(input, ws)
		if chatHistoryCtx != "" {
			return chatHistoryCtx
		}
		return ""
	}

	var context strings.Builder
	context.WriteString("\n# IMPLICIT CONTEXT (Pre-fetched based on your request):\n")

	foundCount := 0
	for _, rec := range records {
		fullPath := rec.Path
		if !filepath.IsAbs(rec.Path) {
			fullPath = filepath.Join(ws.Path, rec.Path)
		}

		data, err := os.ReadFile(fullPath)
		if err == nil {
			foundCount++
			content := string(data)
			if len(content) > 2000 {
				content = content[:2000] + "\n... (truncated)"
			}
			context.WriteString(fmt.Sprintf("\n## File: %s\n```\n%s\n```\n", rec.Path, content))
		}
		
		if foundCount >= 3 {
			break
		}
	}

	if foundCount == 0 {
		// Fallback: search chat history
		chatHistoryCtx := searchChatHistory(input, ws)
		if chatHistoryCtx != "" {
			return chatHistoryCtx
		}
		return ""
	}

	return context.String()
}

// searchChatHistory searches .ruthen/chat_history.md for relevant past conversations
// by scanning for recent QA blocks that match the user's input.
func searchChatHistory(input string, ws *Workspace) string {
	if !ws.Active {
		return ""
	}

	chatPath := filepath.Join(ws.Path, ".ruthen", "chat_history.md")
	data, err := os.ReadFile(chatPath)
	if err != nil {
		return ""
	}

	content := string(data)
	// Split into QA blocks separated by ---
	blocks := strings.Split(content, "---")
	if len(blocks) == 0 {
		return ""
	}

	inputLower := strings.ToLower(input)
	var relevantBlocks []string

	for _, block := range blocks {
		block = strings.TrimSpace(block)
		if block == "" {
			continue
		}

		blockLower := strings.ToLower(block)

		// Check if any significant words from the input appear in this block
		inputWords := strings.Fields(inputLower)
		matchCount := 0
		for _, word := range inputWords {
			if len(word) < 4 {
				continue
			}
			if strings.Contains(blockLower, word) {
				matchCount++
			}
		}

		// A match is relevant if at least 2 significant words match
		if matchCount >= 3 {
			relevantBlocks = append(relevantBlocks, block)
		}
	}

	if len(relevantBlocks) == 0 {
		return ""
	}

	// Return up to 3 relevant blocks
	var result strings.Builder
	result.WriteString("\n# IMPLICIT CONTEXT (Past conversation history):\n")
	for i, block := range relevantBlocks {
		if i >= 3 {
			break
		}
		// Truncate long blocks
		if len(block) > 1000 {
			block = block[:1000] + "\n... (truncated)"
		}
		result.WriteString(fmt.Sprintf("\n### Past Discussion:\n%s\n", block))
	}

	return result.String()
}

File: UNIT-01/orchestrator/src/auto_context.rs
use crate::clients::indexer::IndexerClient;
use crate::workspace::Workspace;
use std::path::Path;

pub async fn get_auto_context(input: &str, ws: &Workspace) -> String {
    if !ws.active {
        return String::new();
    }

    let indexer = IndexerClient::new();
    match indexer.search(input).await {
        Ok(records) if !records.is_empty() => {
            let mut context =
                String::from("\n# IMPLICIT CONTEXT (Pre-fetched based on your request):\n");
            let mut found = 0;
            for rec in &records {
                let full_path = if Path::new(&rec.path).is_absolute() {
                    rec.path.clone()
                } else {
                    format!("{}/{}", ws.path, rec.path)
                };
                if let Ok(data) = std::fs::read_to_string(&full_path) {
                    found += 1;
                    let content = if data.len() > 2000 {
                        format!("{}\n... (truncated)", &data[..2000])
                    } else {
                        data
                    };
                    context.push_str(&format!("\n## File: {}\n```\n{}\n```\n", rec.path, content));
                }
                if found >= 3 {
                    break;
                }
            }
            if found > 0 {
                return context;
            }
        }
        Ok(_) => {}
        Err(_) => {}
    }

    search_chat_history(input, ws)
}

fn search_chat_history(input: &str, ws: &Workspace) -> String {
    if !ws.active {
        return String::new();
    }

    let chat_path = Path::new(&ws.path).join(".ruthen").join("chat_history.md");
    let content = match std::fs::read_to_string(&chat_path) {
        Ok(c) => c,
        Err(_) => return String::new(),
    };

    let input_lower = input.to_lowercase();
    let input_words: Vec<&str> = input_lower
        .split_whitespace()
        .filter(|w| w.len() >= 4)
        .collect();

    if input_words.is_empty() {
        return String::new();
    }

    let mut relevant: Vec<String> = Vec::new();
    for block in content.split("---") {
        let block = block.trim();
        if block.is_empty() {
            continue;
        }
        let block_lower = block.to_lowercase();
        let matches = input_words
            .iter()
            .filter(|w| block_lower.contains(*w))
            .count();
        if matches >= 3 {
            let truncated = if block.len() > 1000 {
                format!("{}\n... (truncated)", &block[..1000])
            } else {
                block.to_string()
            };
            relevant.push(truncated);
            if relevant.len() >= 3 {
                break;
            }
        }
    }

    if relevant.is_empty() {
        return String::new();
    }

    let mut result = String::from("\n# IMPLICIT CONTEXT (Past conversation history):\n");
    for block in &relevant {
        result.push_str(&format!("\n### Past Discussion:\n{}\n", block));
    }
    result
}
